{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ChIP-Seq analysis notebook\n",
    "\n",
    "This interactive computational notebook is a template for analyzing ChIP-seq data preprocessed by BWA, MACS2, MEME, and MAST, performed by a companion upstream pipeline. \n",
    "\n",
    "To get started, put your processed pipeline outputs in the `data` folder.\n",
    "\n",
    "## Setup the environment\n",
    "Set the name of the experiment below. \"E-1\" is the name of the test experiment, using sample data from PPARGC1A ChIP-Seq in HepG2 generated by [Charos et al.](https://www.ncbi.nlm.nih.gov/pubmed/22955979)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "expt_id = 'E-1'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's next load all the necessary libraries, and check that the data can be loaded."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Conflicts with tidy packages ---------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "# Load packages and set up notebook environment\n",
    "options(warn = -1)\n",
    "options(jupyter.plot_mimetypes = 'image/png')\n",
    "suppressPackageStartupMessages(require(tidyverse))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Unzip the data\n",
    "unzip(zipfile, overwrite=TRUE, exdir='data')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load MACS2 peak data\n",
    "fpath = paste0('data/macs2_output/', expt_id, '_peaks.sorted.tsv')\n",
    "peaks = read.csv(fpath, sep='\\t', header=F, \n",
    "                 col.names=c('chrom', 'start', 'end', 'length', 'abs_summit', 'pileup', \n",
    "                             'log10p', 'fold_enrichment', 'log10q', 'name'))\n",
    "head(peaks)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Histogram of peak scores\n",
    "Stuart recommended two histogram plots as quality controls. The first displays the number of peaks by the peak score. Most peaks should have low scores, with a \"long tail\" of high-scoring peaks that represent the biologically relevant binding events.\n",
    "\n",
    "This overall peak score histogram is plotted below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "options(repr.plot.width=6, repr.plot.height=3)\n",
    "ggplot(peaks, aes(x=log10q)) + geom_histogram(binwidth=1) + \n",
    "    ggtitle('Figure 1: Histogram of peak scores') +\n",
    "    xlab('-log10(q-value)') + ylab('number of peaks')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TSS coverage profiles\n",
    "The ChIPseeker package (vignette [here](http://bioconductor.org/packages/release/bioc/vignettes/ChIPseeker/inst/doc/ChIPseeker.html)\n",
    ") was used to produce the ChIP-seq-specific plots below, such as a coverage heatmap relative to transcription start sites (TSS) across the genome, and to perform GO enrichment of nearby genes.\n",
    "\n",
    "Load the packages and read the processed peaks. Peak processing was performed in the pipeline stage, following the steps recommended by Stuart:\n",
    "- Filter blacklisted regions\n",
    "- Filter non-canonical contigs (i.e., not chromosomes 1-22/X/Y/Mt) \n",
    "- Sort by coverage\n",
    "- Extend by 100bp on each side\n",
    "- Take top 300 peaks (by coverage)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "suppressPackageStartupMessages(library('ChIPseeker'))\n",
    "suppressPackageStartupMessages(library('TxDb.Hsapiens.UCSC.hg38.knownGene'))\n",
    "suppressPackageStartupMessages(library('org.Hs.eg.db'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "peak <- readPeakFile('data/macs2_output/E-1_peaks.sorted.filtered.extended.bed')\n",
    "txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, plot the peak coverage as a heatmap centered on promoter locations. This plot provides a granular view of all peak tags within 3kb of a promoter."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "options(repr.plot.width=6, repr.plot.height=6)\n",
    "promoter <- getPromoters(TxDb=txdb, upstream=3000, downstream=3000)\n",
    "tagMatrix <- getTagMatrix(peak, windows=promoter)\n",
    "tagHeatmap(tagMatrix, xlim=c(-3000, 3000), color=\"red\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The tag matrix (tag x position relative to promoter) can also be summarized by total coverage at each position."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "options(repr.plot.height=4, repr.plot.width=6)\n",
    "plotAvgProf(tagMatrix, xlim=c(-3000, 3000),\n",
    "            xlab=\"Genomic Region (5'->3')\", ylab = \"Read Count Frequency\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Annotate genes and gene regions\n",
    "Now that we know that binding sites are centered on promoters, as expected, we can search for patterns of enrichment in the gene annotations for these bound genes. ChIPseeker provides an annotation function using the EntrezGene (i.e., RefSeq) gene model to assign the nearest gene features to each peak."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "peak_annots <- annotatePeak('data/macs2_output/E-1_peaks.sorted.filtered.extended.bed', tssRegion=c(-3000, 3000),\n",
    "                         TxDb=txdb, annoDb=\"org.Hs.eg.db\")\n",
    "\n",
    "peak_annots_df <- as.data.frame(peak_annots)\n",
    "head(peak_annots_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can write this out to comma-separated format for opening with Excel:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "write.csv(peak_annots_df, file='peak_annots.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The raw gene feature annotations can be viewed as a pie chart, showing the percentage of peaks landing on promoters, versus other gene features. As expected, promoters and intergenic regions are the dominant feature since this experiment uses transcription factors. This plot may be useful for characterizing the differences in binding patterns between TF and histones."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "options(repr.plot.height=5, repr.plot.width=7)\n",
    "plotAnnoPie(peak_annots)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The Gene Ontology annotations can be assigned and searched for enrichment in the group of bound genes, compared to their frequency in the genome. The [clusterProfiler](http://bioconductor.org/packages/release/bioc/vignettes/clusterProfiler/inst/doc/clusterProfiler.html)\n",
    "package can generate enrichment scores from a list of Entrez Gene IDs derived from the peak annotation step."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# GO analysis using clusterProfiler\n",
    "suppressPackageStartupMessages(library(clusterProfiler))\n",
    "gene_ids <- as.data.frame(peak_annots)$geneId\n",
    "gene <- bitr(gene_ids, fromType='ENTREZID', toType='SYMBOL', OrgDb='org.Hs.eg.db')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, we extract Biological Process (BP) annotations, which are generally the most informative. We can manually inspect the most frequently occurring annotations, but this will just give us the most general (therefore the most common) annotations, not the ones that are more common than expected."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "ggo <- groupGO(gene     = gene_ids,\n",
    "               OrgDb    = org.Hs.eg.db,\n",
    "               ont      = \"BP\",\n",
    "               level    = 3,\n",
    "               readable = TRUE)\n",
    "cat('Number of annotations')\n",
    "nrow(ggo)\n",
    "cat('Top annotation counts')\n",
    "ggo %>% as.data.frame %>% arrange(desc(Count)) %>% head"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "More informative are enriched annotations. We can use enrichGO to find which of these annotations are statistically significant by the hypergeometric test. Heat shock protein annotations bubble to the top, with highly significant q-values (i.e., Benjamini-Hochberg FDR adjusted p-values)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "ego <- enrichGO(gene          = gene_ids,\n",
    "                OrgDb         = org.Hs.eg.db,\n",
    "                ont           = \"BP\",\n",
    "                pAdjustMethod = \"BH\",\n",
    "                readable      = TRUE)\n",
    "head(ego)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's write these to a `csv` file as well. Check in the `notebooks` directory for this Excel-readable file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "write.csv(ego, file='ego.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MAST motif quantification\n",
    "As part of the results zipfile created by the pipeline, there is an html file of the MAST algorithm output. This algorithm uses the top motifs returned by MEME, and quantifies their occurrences (hits) in the raw alignments. \n",
    "\n",
    "Stuart recommended plotting the distribution of hit counts versus p-value of significance of the hit, for each motif. Here we parse the XML-formatted file provided by MAST, and extract the hit counts for each motif, numbered 0 to 2 (in the order displayed in the MEME report)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "library(XML)\n",
    "mast_xml <- xmlParse('data/mast_output/mast.xml')\n",
    "\n",
    "mast_xml_to_hits <- function(mast_xml) {\n",
    "  data <- xmlToList(mast_xml)\n",
    "  \n",
    "  # Dig into the tree with rbind and lapply to extract a data.frame\n",
    "  df <- \n",
    "    do.call('rbind', lapply(data$sequences, function(seq) {\n",
    "      do.call('rbind', lapply(seq[names(seq) == 'seg'], function(seg) {\n",
    "        do.call('rbind', lapply(seg[names(seg) == 'hit'], function(hit) {\n",
    "          return(data.frame(\n",
    "            motif=as.integer(hit['idx']), \n",
    "            pvalue=as.numeric(hit['pvalue'])\n",
    "          ))\n",
    "        }))\n",
    "      }))\n",
    "    }))\n",
    "  \n",
    "  rownames(df) <- NULL\n",
    "  return(df)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "options(repr.plot.height=4, repr.plot.width=6)\n",
    "df <- mast_xml_to_hits('data/mast_output/mast.xml')\n",
    "ggplot(df, aes(x=-log10(pvalue))) + geom_histogram(binwidth=0.25) + facet_wrap(~motif)"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.4.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
